File: 2024/golang/day4/day4.go
package main

import (
	"fmt"
	"slices"
	"strings"

	"github.com/schwarmco/go-cartesian-product"
)

const PAD = 3
const LINE_LEN = 140

func reverseString(s string) string {
	str_rune := []rune(s)

	for i, j := 0, len(str_rune)-1; i < j; i, j = i+1, j-1 {
		str_rune[i], str_rune[j] = str_rune[j], str_rune[i]
	}

	result := string(str_rune)
	return result
}

func find_num_instances_sequential(substr string, grid [][]string, pos Pair) int {
	left_value := grid[pos.y][pos.x] + grid[pos.y][pos.x-1] + grid[pos.y][pos.x-2] + grid[pos.y][pos.x-3]
	right_value := grid[pos.y][pos.x] + grid[pos.y][pos.x+1] + grid[pos.y][pos.x+2] + grid[pos.y][pos.x+3]
	bottom_value := grid[pos.y][pos.x] + grid[pos.y+1][pos.x] + grid[pos.y+2][pos.x] + grid[pos.y+3][pos.x]
	top_value := grid[pos.y][pos.x] + grid[pos.y-1][pos.x] + grid[pos.y-2][pos.x] + grid[pos.y-3][pos.x]
	top_left_value := grid[pos.y][pos.x] + grid[pos.y-1][pos.x-1] + grid[pos.y-2][pos.x-2] + grid[pos.y-3][pos.x-3]
	top_right_value := grid[pos.y][pos.x] + grid[pos.y-1][pos.x+1] + grid[pos.y-2][pos.x+2] + grid[pos.y-3][pos.x+3]
	bottom_left_value := grid[pos.y][pos.x] + grid[pos.y+1][pos.x-1] + grid[pos.y+2][pos.x-2] + grid[pos.y+3][pos.x-3]
	bottom_right_value := grid[pos.y][pos.x] + grid[pos.y+1][pos.x+1] + grid[pos.y+2][pos.x+2] + grid[pos.y+3][pos.x+3]

	total := 0
	if left_value == substr {
		total += 1
	}
	if right_value == substr {
		total += 1
	}
	if bottom_value == substr {
		total += 1
	}
	if top_value == substr {
		total += 1
	}
	if top_left_value == substr {
		total += 1
	}
	if top_right_value == substr {
		total += 1
	}
	if bottom_left_value == substr {
		total += 1
	}
	if bottom_right_value == substr {
		total += 1
	}

	return total
}

func do_instances_cross(substr string, grid [][]string, pos Pair) bool {
	left_diag_value := grid[pos.y-1][pos.x-1] + grid[pos.y][pos.x] + grid[pos.y+1][pos.x+1]
	right_diag_value := grid[pos.y+1][pos.x-1] + grid[pos.y][pos.x] + grid[pos.y-1][pos.x+1]

	total := 0

	if left_diag_value == substr || left_diag_value == reverseString(substr) {
		total += 1
	}
	if right_diag_value == substr || right_diag_value == reverseString(substr) {
		total += 1
	}

	return total == 2
}

func getCartesian() chan []interface{} {
	x := make([]interface{}, 0)
	for n := range LINE_LEN {
		x = append(x, n+PAD)
	}

	coords := cartesian.Iter(x, x)

	return coords
}

func partOne(input [][]string) {
	coords := getCartesian()

	total := 0
	for product := range coords {
		pair := Pair{x: product[0].(int), y: product[1].(int)}
		total += find_num_instances_sequential("XMAS", input, pair)
	}

	fmt.Println(total)
}

func partTwo(input [][]string) {
	coords := getCartesian()

	total := 0
	for product := range coords {
		pair := Pair{x: product[0].(int), y: product[1].(int)}
		if do_instances_cross("MAS", input, pair) {
			total += 1
		}
	}

	fmt.Println(total)
}

func main() {
	blank_line := slices.Repeat([]string{"-"}, LINE_LEN+(PAD*2))
	blank_lines := slices.Repeat([][]string{blank_line}, (PAD * 2))
	starting_row := PAD
	output := getOutput()
	for _, val := range output {
		next_line := slices.Repeat([]string{"-"}, PAD)
		next_line = append(next_line, strings.Split(val, "")...)
		next_line = append(next_line, slices.Repeat([]string{"-"}, PAD)...)
		blank_lines = slices.Insert(blank_lines, starting_row, next_line)
		starting_row += 1
	}

	partOne(blank_lines)
	partTwo(blank_lines)
}

File: 2024/day4.py
import re
from itertools import product

from utils import get_file_output


PAD = 3
LINE_LEN = 140
output = get_file_output("day4.txt")
lines = [["-"] * (LINE_LEN + (PAD * 2))] * (PAD * 2)
starting_row = PAD
for line in output:
    lines.insert(starting_row, (["-"] * PAD) + list(line.strip()) + (["-"] * PAD))
    starting_row += 1

# lines.extend([(["-"] * (LINE_LEN + (PAD * 2)))] * PAD)


def find_num_instances_sequential(substr: str, grid: list[list], pos: tuple):
    # search left
    left_value = grid[pos[0]][pos[1]] + grid[pos[0]][pos[1] - 1] + grid[pos[0]][pos[1] - 2] + grid[pos[0]][pos[1] - 3]
    # search right
    right_value = grid[pos[0]][pos[1]] + grid[pos[0]][pos[1] + 1] + grid[pos[0]][pos[1] + 2] + grid[pos[0]][pos[1] + 3]
    # search up
    up_value = grid[pos[0]][pos[1]] + grid[pos[0] - 1][pos[1]] + grid[pos[0] - 2][pos[1]] + grid[pos[0] - 3][pos[1]]
    # search down
    down_value = grid[pos[0]][pos[1]] + grid[pos[0] + 1][pos[1]] + grid[pos[0] + 2][pos[1]] + grid[pos[0] + 3][pos[1]]
    # search top left
    top_left_value = grid[pos[0]][pos[1]] + grid[pos[0] - 1][pos[1] - 1] + grid[pos[0] - 2][pos[1] - 2] + grid[pos[0] - 3][pos[1] - 3]
    # search top right
    top_right_value = grid[pos[0]][pos[1]] + grid[pos[0] - 1][pos[1] + 1] + grid[pos[0] - 2][pos[1] + 2] + grid[pos[0] - 3][pos[1] + 3]
    # search bottom left
    bottom_left_value = (grid[pos[0]][pos[1]] + grid[pos[0] + 1][pos[1] - 1] + grid[pos[0] + 2][pos[1] - 2] + grid[pos[0] + 3][pos[1] - 3])  # noqa
    # search bottom right
    bottom_right_value = (grid[pos[0]][pos[1]] + grid[pos[0] + 1][pos[1] + 1] + grid[pos[0] + 2][pos[1] + 2] + grid[pos[0] + 3][pos[1] + 3])  # noqa

    values = {
        "left_value": 1 if left_value == substr else 0,
        "right_value": 1 if right_value == substr else 0,
        "up_value": 1 if up_value == substr else 0,
        "down_value": 1 if down_value == substr else 0,
        "bottom_left_value": 1 if bottom_left_value == substr else 0,
        "bottom_right_value": 1 if bottom_right_value == substr else 0,
        "top_left_value": 1 if top_left_value == substr else 0,
        "top_right_value": 1 if top_right_value == substr else 0,
    }

    return sum(values.values())


def find_num_instances_crossed(substr: str, grid: list[list], pos: tuple):
    left_diag_coords = [(pos[0] - 1, pos[1] - 1), (pos[0], pos[1]), (pos[0] + 1, pos[1] + 1)]
    right_diag_coords = [(pos[0] + 1, pos[1] - 1), (pos[0], pos[1]), (pos[0] - 1, pos[1] + 1)]
    left_diag = "".join([grid[x][y] for x,y in left_diag_coords])
    right_diag = "".join([grid[x][y] for x,y in right_diag_coords])

    values = {
        "left_diag": 1 if (left_diag == substr or left_diag == substr[::-1]) else 0,
        "right_diag": 1 if (right_diag == substr or right_diag == substr[::-1]) else 0,
    }

    return sum(values.values()) == 2


def part_one():
    total = 0
    for x, y in product(range(PAD, LINE_LEN + PAD), range(PAD, LINE_LEN + PAD)):
       total += find_num_instances_sequential("XMAS", lines, (x, y))

    print(total)


def part_two():
    total = 0
    for x, y in product(range(PAD, LINE_LEN + PAD), range(PAD, LINE_LEN + PAD)):
       total += 1 if find_num_instances_crossed("MAS", lines, (x, y)) else 0

    print(total)



if __name__ == '__main__':
    part_one()
    part_two()
